safe_transaction_service/tokens/tasks.py

```diff
@@ -63,36 +63,36 @@ def calculate_token_eth_price_task(
     )  # Expire in 15 minutes
     if key_was_set or force_recalculation:
         price_service = PriceServiceProvider()
-        eth_price = (
-            price_service.get_token_eth_value(token_address)
-            or price_service.get_token_usd_price(token_address)
-            / price_service.get_native_coin_usd_price()
-        )
-        if not eth_price:  # Try composed oracles
-            if underlying_tokens := price_service.get_underlying_tokens(token_address):
-                eth_price = 0
-                for underlying_token in underlying_tokens:
-                    # Find underlying token price and multiply by quantity
-                    address = underlying_token.address
-                    eth_price += (
-                        calculate_token_eth_price_task(
-                            address,
-                            f"price-service:{address}:eth-price",  # TODO Refactor all the calculation logic
-                        ).eth_value
-                        * underlying_token.quantity
-                    )
-        if eth_price:
-            eth_value_with_timestamp = EthValueWithTimestamp(eth_price, now)
-            redis.setex(redis_key, redis_expiration_time, str(eth_value_with_timestamp))
-            if not getattr(settings, "CELERY_ALWAYS_EAGER", False):
-                # Recalculate price before cache expires and prevents recursion checking Celery Eager property
-                calculate_token_eth_price_task.apply_async(
-                    (token_address, redis_key),
-                    {"force_recalculation": True},
-                    countdown=redis_expiration_time - 300,
-                )
-        else:
-            logger.warning("Cannot calculate eth price for token=%s", token_address)
+        eth_price = price_service.get_eth_price_oracles(token_address)
+        if not eth_price:
+            eth_price = price_service.get_eth_price_composed_oracles(token_address)
+
+        logger.debug("Calculated eth-price=%f for token=%s", eth_price, token_address)
+        if not eth_price:
+            logger.warning(
+                "Cannot calculate eth price for token=%s - Trying to use previous price",
+                token_address,
+            )
+            last_redis_value = redis.get(redis_key)
+            if last_redis_value:
+                logger.warning("Using previous eth price for token=%s", token_address)
+                eth_price = EthValueWithTimestamp.from_string(
+                    last_redis_value.decode()
+                ).eth_value
+            else:
+                logger.warning("Cannot calculate eth price for token=%s", token_address)
```

Comment:
Could `eth_price` still being `0.0` at this point? It could happend in some (rare) scenarios, right? In that case I assume we prefer sending `0.0` to the client and don't throw an exception.

Reply:
This is currently returning `0.0` and don't throwing an exception :) `logger.warning` is a log warning level information.

Reply:
Yes, I know, I mean, we chose returning 0.0 over throwing an exception, right? My doubt is if it could be better to actually throw the exception 🙂

Reply:
Ok, I read some old discussions about this and the conclusion was that 0 is an adequate value in these cases, so nothing to argue on my side, great job @moisses89 🙂

Reply:
If you return an exception here there will never be a price for the token and calculation will be retried

Reply:
@uxio0 I'm not sure if we want this, could increase a lot the tasks for tokens that have not price.

Reply:
Maybe I was not clear, we don't want to raise an exception 😉